test/functional/test_framework/ellsq.py

```diff
@@ -0,0 +1,243 @@
+    [(fe(0x49a0dc068c3f117aefdc842d3d358153f677f04c6dabc9c91b09d452fef27b66), fe(0x7b944da48a175dbc444ead8db82eff66b081a8aae6453fed2bca9720b44dd6e5)), [fe(0x7bf1e2b1720c1c440db64687f16439fa41b398338095f24ebeec0cfa88750dc9), fe(0xdc97e26d3137445d6c1269b61a7655010c19c36a2e361066e31e2bb10403470b), None                                                                  , None                                                                  ]],
+    [(fe(0xd09a4047f158fe52f96c661d02c68657c4c976ea96ea85ef46d6985bd540756b), fe(0xe793bfaae9300f18e6f9b55aae26322368b61d51ae5022efe266c72d574178bc)), [fe(0x7e6175fdfbb9fb4faf6e2b925ef86c4a444d819aaa82dbee545d3d9b296375be), None                                                                  , None                                                                  , None                                                                  ]],
+    [(fe(0x3498662504b73c7c8cecb6c33cd493bdfc190e0f87d913d7ff9ad42e222bfe95), fe(0x245b3a61b8d46997f14f2fea2874899691eb32542b9907d65eb9d21d42454021)), [fe(0x7f556282c3dd9d263390d6bbddada698ab8fd7c7d1a06498f42b30437c8361ad), None                                                                  , None                                                                  , None                                                                  ]]
+]
+
+def encode(P):
+    while True:
+        u = fe(random.randrange(1, SECP256K1_ORDER))
+        fe1 = f(u)
+        # convert fe1 to jacobian form for EC operations
+        fe1 = (fe1[0].val, fe1[1].val, 1)
+        T = SECP256K1.negate(fe1)
+        Q = SECP256K1.add(T, P)
+        if SECP256K1.is_infinity(Q):
+            Q = T
+        j = secrets.choice([0,1,2,3])
+        Q = SECP256K1.affine(Q)
+        v = r(fe(Q[0]), fe(Q[1]), j)
+        if v is not None:
+            return (u, v)
+
+def decode(u, v):
+    fe1 = f(u)
+    fe2 = f(v)
+    # convert fe1 and fe2 to jacobian form for EC operations
+    jac1 = (fe1[0].val, fe1[1].val, 1)
+    jac2 = (fe2[0].val, fe2[1].val, 1)
+    T = jac1
+    S = jac2
+    P = SECP256K1.add(T, S)
+    if SECP256K1.is_infinity(P):
+        P = T
+    P = SECP256K1.affine(P)
+    return (fe(P[0]), fe(P[1]))
+
+FIELD_BITS = SECP256K1_FIELD_SIZE.bit_length()
+FIELD_BYTES = (FIELD_BITS + 7) // 8
+
+def encode_bytes(P):
+    u, v = encode(P)
+    up = u.val
+    vp = v.val
+    return up.to_bytes(FIELD_BYTES, 'big') + vp.to_bytes(FIELD_BYTES, 'big')
+
+def decode_bytes(enc):
+    u = (int.from_bytes(enc[:FIELD_BYTES], 'big') & ((1 << FIELD_BITS) - 1)) % SECP256K1_FIELD_SIZE
```

Comment:
The `& ((1 << FIELD_BITS) - 1)` doesn't do anything here, as FIELD_BITS is 256, an exact multiple of 8.

Similar to above, what do you think about adding a `from_bytes(b)` method to fe (returning `fe(int.from_bytes(b, 'big'))`), and then using `return decode(fe.from_bytes(enc[:32]), fe.from_bytes(enc[32:]))` here?

Reply:
Thanks! Done.

Reply:
Thanks! Done.